netty/src/main/java/io/grpc/netty/NettyServer.java

```diff
@@ -258,51 +292,88 @@ public void operationComplete(ChannelFuture future) throws Exception {
         ch.closeFuture().addListener(loopReleaser);
       }
     });
-    // Bind and start to accept incoming connections.
-    ChannelFuture future = b.bind(address);
-    // We'd love to observe interruption, but if interrupted we will need to close the channel,
-    // which itself would need an await() to guarantee the port is not used when the method returns.
-    // See #6850
-    future.awaitUninterruptibly();
-    if (!future.isSuccess()) {
-      throw new IOException(String.format("Failed to bind to address %s", address), future.cause());
+
+    Future<Map<ChannelFuture, SocketAddress>> bindCallFuture =
+        bossExecutor.submit(new Callable<Map<ChannelFuture, SocketAddress>>() {
+          @Override
+          public Map<ChannelFuture, SocketAddress> call() {
+            Map<ChannelFuture, SocketAddress> map = new HashMap<>();
+            for (SocketAddress address: addresses) {
+                ChannelFuture future = b.bind(address);
+                map.put(future, address);
+            }
+            return map;
+          }
+        }
+    );
+    Map<ChannelFuture, SocketAddress> channelFutures;
+    try {
+      Future<Map<ChannelFuture, SocketAddress>> bindCallResult =
+          bindCallFuture.awaitUninterruptibly();
+      if (!bindCallResult.isSuccess()) {
+        throw bindCallResult.cause();
+      }
+      channelFutures = bindCallResult.get();
+    } catch (Throwable ex) {
```

Comment:
Catching Throwable is generally wrong, unless it is being re-thrown unchanged. Certainly here where we wrap it in an IOException is probably a bad idea. Errors should generally propagate.

I think the `bindCallResult.cause()` should just be wrapped in an IOException immediately. The `get()` should be replaced with `getNow()` to avoid potential exceptions. At that point, there's no need for the catch.

Reply:
Thanks. It made it simpler and more readable now. Still we throw IOException for whatever failures that prevent us from getting the bindfuture, although it may not be from the bind itself. I guess it is somewhat unnecessary to throw different exceptions types at a higher granularity, at least for now.